pallets/transfer-allowlist/src/lib.rs

```diff
@@ -0,0 +1,481 @@
+		/// In that case it is set to the current block + delay.
+		/// if the allowance is later updated with the add allowance extrinsic, it is set back to max.
+		pub blocked_at: BlockNumber,
+	}
+
+	impl<BlockNumber> AllowanceDetails<BlockNumber>
+	where
+		BlockNumber: AtLeast32BitUnsigned,
+	{
+		fn default() -> Self {
+			Self {
+				allowed_at: BlockNumber::zero(),
+				blocked_at: BlockNumber::max_value(),
+			}
+		}
+	}
+
+	/// Storage item for whether a sending account and currency have restrictions set
+	/// a double map is used here as we need to know whether there is a restriction set
+	/// for the account and currency.
+	/// Using an StorageNMap would not allow us to look up whether there was a restriction for the sending account and currency, given that:
+	/// - we're checking whether there's an allowance specified for the receiver location
+	///   - we would only find whether a restriction was set for the account in this caseif:
+	///     - an allowance was specified for the receiving location, which would render blocked restrictions useless
+	/// - we would otherwise need to store a vec of locations, which is problematic given that there isn't a set limit on receivers
+	/// If a transfer restriction is in place, then a second lookup is done on
+	/// AccountCurrencyAllowances to see if there is an allowance for the reciever
+	/// This allows us to keep storage map vals to known/bounded sizes.
+	#[pallet::storage]
+	#[pallet::getter(fn sender_currency_restriction_set)]
+	pub type AccountCurrencyTransferRestriction<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat,
+		T::AccountId,
+		Twox64Concat,
+		T::CurrencyId,
+		u64,
+		OptionQuery,
+	>;
+
+	/// Storage item for allowances specified for a sending account, currency type and recieving location
+	#[pallet::storage]
+	#[pallet::getter(fn sender_currency_reciever_allowance)]
+	pub type AccountCurrencyTransferAllowance<T: Config> = StorageNMap<
+		_,
+		(
+			NMapKey<Twox64Concat, T::AccountId>,
+			NMapKey<Twox64Concat, T::CurrencyId>,
+			NMapKey<Blake2_128Concat, Location>,
+		),
+		AllowanceDetails<T::BlockNumber>,
+		OptionQuery,
+	>;
+	/// Storage item for Allowance delays for a sending account/currency
+	#[pallet::storage]
+	#[pallet::getter(fn sender_currency_delay)]
+	pub type AccountCurrencyDelay<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat,
+		T::AccountId,
+		Twox64Concat,
+		T::CurrencyId,
+		T::BlockNumber,
+		OptionQuery,
+	>;
+
+	//
+	// Pallet Errors and Events
+	//
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Number of allowances for a sending Account/Currency set at max allowance count storage type val (currently u64::MAX)
+		AllowanceCountOverflow,
+		/// An operation expecting one or more allowances for a sending Account/Currency set, where none present
+		NoAllowancesSet,
+		/// Attempted to create allowance for existing Sending Account, Currency, and Receiver combination
+		ConflictingAllowanceSet,
```

Comment:
From the /// comment, I guess this means that this allowance already exists? If so, let's make it consistent with similar error names:

```suggestion
		DuplicateAllowance,
```

Reply:
👍 fixed